src/failure_detector.rs

```diff
@@ -0,0 +1,373 @@
+            size,
+            is_filled: false,
+            index: 0,
+            sum: 0.0,
+            mean: 0.0,
+        }
+    }
+
+    /// Calculates the mean
+    pub fn mean(&self) -> f64 {
+        self.mean
+    }
+
+    /// Appends a new value.
+    pub fn append(&mut self, interval: f64) {
+        if self.index == self.size {
+            self.is_filled = true;
+            self.index = 0;
+        }
+
+        if self.is_filled {
+            self.sum -= self.data[self.index];
+        }
+        self.sum += interval;
+
+        self.data[self.index] = interval;
+        self.index += 1;
+
+        self.mean = self.sum / self.len() as f64;
+    }
+
+    fn len(&self) -> usize {
+        if self.is_filled {
+            return self.size;
+        }
+        self.index
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use mock_instant::MockClock;
+    use rand::prelude::*;
+
+    use super::{BoundedArrayStats, SamplingWindow};
+    use crate::failure_detector::{FailureDetector, FailureDetectorConfig};
+
+    #[test]
+    fn test_failure_detector() {
+        let mut rng = rand::thread_rng();
+        let mut failure_detector = FailureDetector::new(FailureDetectorConfig::default());
+
+        let intervals_options = vec![1u64, 2];
+        let node_ids_options = vec!["node-1", "node-2", "node-3"];
+        for _ in 0..=2000 {
+            let time_offset = intervals_options.choose(&mut rng).unwrap();
+            let node_id = node_ids_options.choose(&mut rng).unwrap();
+            MockClock::advance(Duration::from_secs(*time_offset));
+            failure_detector.report_heartbeat(node_id);
+        }
+
+        for node_id in &node_ids_options {
+            failure_detector.update_node_liveliness(node_id);
+        }
+
+        let mut lives = failure_detector.live_nodes().collect::<Vec<_>>();
+        lives.sort();
+
+        assert_eq!(lives, vec!["node-1", "node-2", "node-3"]);
+        assert_eq!(
+            failure_detector.dead_nodes().collect::<Vec<&str>>(),
+            Vec::<&str>::new()
+        );
+    }
+
+    #[test]
+    fn test_failure_detector_node_down() {
+        let mut rng = rand::thread_rng();
+        let mut failure_detector = FailureDetector::new(FailureDetectorConfig::default());
+        let intervals_options = vec![1u64, 2];
+
+        for _ in 0..=2000 {
+            let time_offset = intervals_options.choose(&mut rng).unwrap();
+            MockClock::advance(Duration::from_secs(*time_offset));
+            failure_detector.report_heartbeat("node-1");
+        }
+
+        failure_detector.update_node_liveliness("node-1");
+        assert_eq!(
+            failure_detector.live_nodes().collect::<Vec<&str>>(),
+            vec!["node-1"]
+        );
+        assert_eq!(
+            failure_detector.dead_nodes().collect::<Vec<&str>>(),
+            Vec::<&str>::new()
+        );
+
+        // stop reporting node-3 heartbeat
+        for _ in 0..=2000 {
```

Comment:
Can't we just advance the clock with a value that's so large that phi becomes > phi_threshold?